### ISSUE TITLE
JWT Support for Kafka output

### ISSUE THREAD

**Issue by chinmaychandak** [elastic/beats#22603]
I would like to scrape logs using Beats (either Filebeat, Metricbeat, etc.) and send them to Logstash or Kafka depending on my use case. Currently, I am leveraging the SSL support, but I would like to use JSON Web Token instead; i.e., the beats should talk to an Authorization server, receive a JWT and then pass it as part of the Auth header bearer token schema when writing logs to an output sink which then validates the JWT. I am specifically looking to integrate JWT in concert with OAuth for my use case (because of needing different ACLs/scopes for operations and data, and also the requirement for making the tokens ephemeral with some expiration policy).

Is this feature available/being worked on? If not, I would like to contribute, and was hoping if someone could give me pointers on where to start? I also think I would need to implement this in the libbeat/common/transport directory just like TLS/SSL?

NOTE: I am completely new to the Beats code base, even though I've used beats quite a bit.

Any help would be greatly appreciated! 

P.S. I had created a topic on the forum first: https://discuss.elastic.co/t/jwt-support-in-beats/254825, but thought it would be helpful to open an issue here since I did not receive many responses there.

**Comment by andrewkroh** [comment]
To clarify, you want to add an OAuth 2.0 capability to the Kafka output? (Logstash doesn't support using anything other than mTLS to secure the connection.) Probably the kerberos support would be a good example to look at libbeat/common/transport/kerberos w.r.t. the outputs. Also Filebeat has some code in its httpjson input for doing oauth so it would be good to stay consistent w.r.t. [configuration](https://github.com/elastic/beats/blob/df102bd5c66fbd2a9effdf63ba9fd77b4b670de6/x-pack/filebeat/input/httpjson/config_oauth.go#L42-L50) options.

**Comment by chinmaychandak** [comment]
Thank you so much for responding, @andrewkroh! 

> you want to add an OAuth 2.0 capability to the Kafka output?

Correct, yes. 

> Probably the kerberos support would be a good example to look at libbeat/common/transport/kerberos w.r.t. the outputs. Also Filebeat has some code in its httpjson input for doing oauth so it would be good to stay consistent w.r.t. configuration options.

Thank you, will take a look at both of these!

**Comment by kfox1111** [comment]
I'm interested in JWT support as well, but the JWT's can be generated on the host automatically via SPIFFE/SPIRE.
So I think the needed support would look similar to the username/password config, just that it passes the string a little differently, and reads it from a file.

**Comment by kfox1111** [comment]
Still would be useful. then could use k8s projected tokens for auth.

**Comment by jmbass** [comment]
👍

**Comment by kfox1111** [comment]
:+1

**Comment by rwaweber** [comment]
👍🏻 

But also, it would be great if beats could support communication with an elasticsearch cluster by JWT as well, especially given [elasticsearch itself supports JWT Authentication](https://www.elastic.co/docs/deploy-manage/users-roles/cluster-or-deployment-auth/jwt), but the only authentication mechanisms available within beats for an elasticsearch cluster seem to be api_key(static credential, non-ideal), username/password, kerberos, and tls client certs.

Part of me feels like this may constitute a second issue given the intent of the initial post here, but I'll defer to you all for how you'd want to track it!

**Comment by andrewkroh** [comment]
> Part of me feels like this may constitute a second issue given the intent of the initial post here

Agree.

> But also, it would be great if beats could support communication with an elasticsearch cluster by JWT as well,

It should already be possible by setting `output.elasticsearch.headers`: 

```yaml
output.elasticsearch.headers:
  Authorization: 'Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJpc3M4IiwiYXVkIjoiYXVkOCIsInN1YiI6InNlY3VyaXR5X3Rlc3RfdXNlciIsImV4cCI6NDA3MDkwODgwMCwiaWF0Ijo5NDY2ODQ4MDB9.UnnFmsoFKfNmKMsVoDQmKI_3-j95PCaKdgqqau3jPMY'
```

Reference:

https://www.elastic.co/docs/reference/beats/filebeat/elasticsearch-output#_headers

**Comment by rwaweber** [comment]
> It should already be possible by setting output.elasticsearch.headers:

True! Though IME many JWT implementations tend to have very short-lived(hours/days, etc.) expiration periods, so instead of asking end users to update their beats configurations constantly, I think it may be substantially more useful to have the elasticsearch output perform the token refresh step on the beats side - as the client-id and client-secret used in that flow tend to be valid for  longer(months/1 year).

Thinking about it a little bit and doing a little perusing of the codebase here, maybe it makes sense to treat it as an OAuth2 transport mechanism, somewhat similar to kerberos? Happy to continue the discussion over in #46662 so I dont distract too much from the original request for [OAuth+Kafka](https://docs.confluent.io/platform/current/security/authentication/oauth-oidc/service-to-service.html) support.

**Comment by cmacknz** [comment]
Supporting this in agent is progress, but the way it's done doesn't cover beats (yet). Reopening.
